### DISCUSSION TITLE
Is it possible to place Markdown footnotes somewhere else in the page?

### DISCUSSION THREAD

**Discussion by pftnhr** [datenstrom/community#921] [Ask a question] [answered]
My posts are structured as follows

``` html
<section>
    <article>
        <header>
        </header>
        <p>Content goes here</p>
        <!-- if there are footnotes -->
        <footer>
           Footnotes...
        </footer>
    </article>
</section>
```
But since I have my [lastedit](https://github.com/pftnhr/yellow-lastedit) extension active, which is inserted via `getExtraHtml()`, it happens with articles with footnotes that the output of `<?php echo $this->yellow->page->getExtraHtml("lastedit") ?>` slips below the `<footer></footer>` area, which not only looks unattractive, but also contradicts the definition of a footer.

My question now is whether I can get the footnotes to the very end of the content or whether I can alternatively hook the output of `<?php echo $this->yellow->page->getExtraHtml("lastedit") ?>` (or other elements that I add in the layout file) into the content of the `*.md` without using JavaScript to manipulate the DOM?

At the moment I have solved this using JavaScript, but I would like to use as little JS as possible (preferably none at all).

**Comment by schulle4u** [comment]
I think you can achieve this using `onParsePageOutput()`. For an usage example have a look at [the meta extension](https://github.com/annaesvensson/yellow-meta), it uses `onParsePageOutput` to insert additional attributes into the opening `<html>` element.

**Reply by pftnhr** [reply]
Thank you for the example! But since I'm not good at regular expressions, I'll leave it at the JavaScript solution for now.

**Reply by pftnhr** [reply]
I played around with `onParsePageOutput()` and it works really well. Thanks to the HTML parser `DOMDocument` it is almost the same as JavaScript.

Unfortunately, the changed output is not displayed in the newsfeed. 😕

**Comment by markseuffert** [comment]
This is uncharted territory as far as I know. Usually the [Markdown parser](https://github.com/annaesvensson/yellow-markdown) takes one piece of raw content data and returns one piece of HTML content data. Markdown footnotes are placed at the bottom of the HTML content data. I like to mention that Markdown footnotes were introduced by the Markdown Extra syntax, they were not part of the original Markdown syntax.

 If someone wants to manipulate Markdown footnotes, here some thoughts and ideas:

* You could put what you want to see in the content file. Put everything that should appear before the footnotes in the content file, for example your `[lastedit]` shortcut. Then you get one big chunk of HTML content data for your layout files. The method `page->getContentHtml()` gives you access to this big chunk of HTML content data. 
* You could break with traditional placement and show footnotes at a specific position on the page. Probably would mean some changes to the Markdown Extra syntax. Then it would work more like a table of contents. You either have to place a `[footnotes]` shortcut in the content file or a `page->getExtraHtml("footnotes")` in the layout file.
* The [API for developers](https://datenstrom.se/yellow/help/api-for-developers) has various events where you could hook into. For example `onParseContentHtml()` lets you modify the HTML output of the Markdown parser. Not sure how much work this would be for a developer. In any case the Yellow API should have everything you need for manipulating Markdown footnotes on the server side. No JavaScript required.

Happy coding. Let me know what you find out.

**Reply by pftnhr** [reply]
Hi Mark, as mentioned above I got it to work with the HTML parser `DOMDocument` and `onParsePageOutput` and in contrast to the JavaScript solution the HTML source code then also matches the output. I have inserted the following code into my theme.php:

```php
     public function onParsePageOutput($page, $text) {
        $output = null;
        if ($this->yellow->page->get("layout") == "blog") {
            $doc = new DOMDocument;
            libxml_use_internal_errors(true);
            $doc->loadHTML($text);
            libxml_clear_errors();

            // Processing begins here
            $footnotesElements = $doc->getElementsByTagName('div');

            foreach ($footnotesElements as $footnotesEnd) {
                if ($footnotesEnd->getAttribute('class') == 'footnotes') {
                    // Remove the hr element
                    $hrElement = $footnotesEnd->getElementsByTagName('hr')->item(0);
                    if ($hrElement) {
                        $footnotesEnd->removeChild($hrElement);
                    }

                    // Create the footer element
                    $footnoteFooter = $doc->createElement("footer");

                    // Move the content of footnotesEnd to the footer element
                    while ($footnotesEnd->firstChild) {
                        $footnoteFooter->appendChild($footnotesEnd->firstChild);
                    }

                    // Add the footer element to the end of the parentElement
                    $footnotesEnd->parentNode->appendChild($footnoteFooter);
                }
            }

            $output = $doc->saveHTML();
        }
        return $output;
    }
```

Unfortunately, I have not found any information on `onParseContentHtml()`. If you have some easy to understand examples, I would be very grateful.

**Reply by markseuffert** [reply]
Have a look at [typography extension](https://github.com/GiovanniSalmeri/yellow-typography) and [toc extension](https://github.com/annaesvensson/yellow-toc), they modify the HTML content data.

**Comment by markseuffert** [comment, marked as answer]
No, this is not possible at the moment.
